File: MiniDOMOffice/CRM/Chat/CChatUser.cs
using System;
using DMD;
using DMD.XML;
using minidom;
using static minidom.Sistema;


namespace minidom
{
    public partial class Messenger
    {

        /// <summary>
        /// Utente appartenente ad una chat
        /// </summary>
        [Serializable]
        public class CChatUser 
            : DMD.XML.DMDBaseXMLObject, IComparable, IComparable<CChatUser>
        {
            /// <summary>
            /// ID dell'utente di sistema
            /// </summary>
            public int uID;

            /// <summary>
            /// Icona associata
            /// </summary>
            public string IconURL;

            /// <summary>
            /// Nome dell'utente di sistema
            /// </summary>
            public string UserName;

            /// <summary>
            /// Nome visibile dell'utente della chat
            /// </summary>
            public string DisplayName;

            /// <summary>
            /// Restituisce o imposta un valore boolano che indica se l'utente é online
            /// </summary>
            public bool IsOnline;

            /// <summary>
            /// Ultima data di accesso
            /// </summary>
            public DateTime? UltimoAccesso;

            /// <summary>
            /// Numero di messaggi non letti
            /// </summary>
            public int MessaggiNonLetti;

            /// <summary>
            /// Costruttore
            /// </summary>
            public CChatUser()
            {
                uID = 0;
                IconURL = "";
                UserName = "";
                DisplayName = "";
                IsOnline = false;
                UltimoAccesso = default;
                MessaggiNonLetti = 0;
            }

            /// <summary>
            /// Serializzazione xml
            /// </summary>
            /// <param name="writer"></param>
            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("UserName", UserName);
                writer.WriteAttribute("DisplayName", DisplayName);
                writer.WriteAttribute("IsOnline", IsOnline);
                writer.WriteAttribute("uID", uID);
                writer.WriteAttribute("IconURL", IconURL);
                writer.WriteAttribute("UltimoAccesso", UltimoAccesso);
                writer.WriteAttribute("MessaggiNonLetti", MessaggiNonLetti);
                base.XMLSerialize(writer);
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "uID":
                        {
                            uID = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "UserName":
                        {
                            UserName = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "DisplayName":
                        {
                            DisplayName = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "IsOnline":
                        {
                            IsOnline = DMD.XML.Utils.Serializer.DeserializeBoolean(fieldValue) == true;
                            break;
                        }

                    case "IconURL":
                        {
                            IconURL = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "UltimoAccesso":
                        {
                            UltimoAccesso = DMD.XML.Utils.Serializer.DeserializeDate(fieldValue);
                            break;
                        }

                    case "MessaggiNonLetti":
                        {
                            MessaggiNonLetti = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }
                    default:
                        base.SetFieldInternal(fieldName, fieldValue);
                        break;
                }
            }


            /// <summary>
            /// Restituisce una stringa che rappresenta l'oggetto
            /// </summary>
            /// <returns></returns>
            public override string ToString()
            {
                return this.UserName;
            }

            /// <summary>
            /// Restituisce il codice hash dell'oggetto
            /// </summary>
            /// <returns></returns>
            public override int GetHashCode()
            {
                return HashCalculator.Calculate(this.UserName);
            }

            /// <summary>
            /// Restituisce true se i due oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public sealed override bool Equals(DMDBaseXMLObject obj)
            {
                return (obj is CChatRoomUser) && this.Equals((CChatRoomUser)obj);
            }

            /// <summary>
            /// Restituisce true se i due oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public virtual bool Euqlas(CChatRoomUser obj)
            {
                return base.Equals(obj)
                    && DMD.Integers.EQ(this.uID, obj.uID)
                    && DMD.Strings.EQ(this.IconURL, obj.IconURL)
                    && DMD.Strings.EQ(this.UserName, obj.UserName)
                    && DMD.Strings.EQ(this.DisplayName, obj.DisplayName)
                    && DMD.Booleans.EQ(this.IsOnline, obj.IsOnline)
                    && DMD.DateUtils.EQ(this.UltimoAccesso, obj.UltimoAccesso)
                    && DMD.Integers.EQ(this.MessaggiNonLetti, obj.MessaggiNonLetti)
                    ;
            }

            /// <summary>
            /// Compara due oggetti in base al nome utente
            /// </summary>
            /// <param name="other"></param>
            /// <returns></returns>
            public int CompareTo(CChatUser other)
            {
                return Strings.Compare(this.UserName, other.UserName, true);
            }

            int IComparable.CompareTo(object obj)
            {
                return this.CompareTo((CChatRoomUser)obj);
            }
        }
    }
}

File: MiniDOMOffice/CRM/Chat/CChatUser.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases

Partial Public Class Messenger

    <Serializable>
    Public Class CChatUser
        Implements XML.IDMDXMLSerializable

        Public uID As Integer
        Public IconURL As String
        Public UserName As String
        Public DisplayName As String
        Public IsOnline As Boolean
        Public UltimoAccesso As Date?
        Public MessaggiNonLetti As Integer

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
            Me.uID = 0
            Me.IconURL = ""
            Me.UserName = ""
            Me.DisplayName = ""
            Me.IsOnline = False
            Me.UltimoAccesso = Nothing
            Me.MessaggiNonLetti = 0
        End Sub

        Protected Sub XMLSerialize(ByVal writer As minidom.XML.XMLWriter) Implements XML.IDMDXMLSerializable.XMLSerialize
            writer.WriteAttribute("UserName", Me.UserName)
            writer.WriteAttribute("DisplayName", Me.DisplayName)
            writer.WriteAttribute("IsOnline", Me.IsOnline)
            writer.WriteAttribute("uID", Me.uID)
            writer.WriteAttribute("IconURL", Me.IconURL)
            writer.WriteAttribute("UltimoAccesso", Me.UltimoAccesso)
            writer.WriteAttribute("MessaggiNonLetti", Me.MessaggiNonLetti)
        End Sub

        Protected Sub SetFieldInternal(ByVal fieldName As String, ByVal fieldValue As Object) Implements XML.IDMDXMLSerializable.SetFieldInternal
            Select Case fieldName
                Case "uID" : Me.uID = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "UserName" : Me.UserName = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "DisplayName" : Me.DisplayName = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "IsOnline" : Me.IsOnline = XML.Utils.Serializer.DeserializeBoolean(fieldValue)
                Case "IconURL" : Me.IconURL = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "UltimoAccesso" : Me.UltimoAccesso = XML.Utils.Serializer.DeserializeDate(fieldValue)
                Case "MessaggiNonLetti" : Me.MessaggiNonLetti = XML.Utils.Serializer.DeserializeInteger(fieldValue)
            End Select
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class

End Class
 